File: demo_create_objc/VPKitDemo/ViewController.m
//
//  ViewController.m
//  VPKitDemo
//
//  Created by jonathan on 01/03/2016.
//  Copyright © 2016 jonathan. All rights reserved.
//

#import "ViewController.h"
#import "ViewController_AutoLayout.h"
#import <VPKit/VPKit.h>

@import AVFoundation;
@import AVKit;

#define LOG_ME 1

@interface ViewController ()
<
  VPKVeepViewerDelegate
, VPKVeepEditorDelegate
, VPKPreviewDelegate
>

@property (nonatomic, strong) VPKVeepViewer* vpViewer;
@property (nonatomic, strong) VPKVeepEditor* vpEditor;


@end

@implementation ViewController




#pragma mark - viewController lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    self.versionLabel.text = [VPKit sdkVersion];
    [self configureViewerWithTestVideo];
    [self configureEditor];
    self.constraints = [[NSMutableArray alloc] init];
    [self configureConstraints];

}

- (BOOL)shouldAutorotate {
    return NO;
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations {
    return UIInterfaceOrientationMaskPortrait;
}

#pragma mark - configuration



- (void)configureViewerWithTestImage {
    UIImage* image = [UIImage imageNamed:@"KrispyGlas"];
    image = [[VPKImage alloc] initWithImage:image veepId:@"1787"];
    self.viewerPreview.image = image;
    self.viewerPreview.delegate = self;

}

- (void)configureViewerWithTestVideo {
    UIImage* image = [UIImage imageNamed:@"tomcruise"];
    image = [[VPKImage alloc] initWithImage:image veepId:@"1788"];
    self.viewerPreview.image = image;
    self.viewerPreview.delegate = self;

    
}


- (void)configureEditor {
    UIImage* image = [UIImage imageNamed:@"stock_photo"];
    image = [[VPKImage alloc] initWithImage:image veepId:nil];
    self.editorPreview.image = image;
    /*
     for the editor example, we'll set an optional delegate.
     */
    self.editorPreview.delegate = self;
}




#pragma mark - VPKPreview delegate


- (void)vpkPreviewTouched:(VPKPreview *)preview image:(VPKImage*)image {
    [preview hideIcon];

    if ([preview isEqual:self.viewerPreview]) {
        [self invokeViewer:image fromView:preview];
    } else {
        [self invokeEditor:image fromView:preview];
    }
}

- (void)invokeViewer:(VPKImage*)image fromView:(UIView*)view {
    
    /*
     invoking the VPKVeepViewer
     
     set the viewer's transitioning delegate to a custom transitioning object (or nil) to override supplied transition animations
     
     this code is all OPTIONAL - if you don't set the delegate on VPKPreview, presenting and dismissing behaviour occurs as a default
     
     */
    
    
    self.vpViewer =  [VPKit viewerWithImage:image
                                   fromView:view];
    self.vpViewer.delegate = self;
    self.vpViewer.modalPresentationStyle = UIModalPresentationOverFullScreen;
    [self presentViewController:self.vpViewer animated:YES completion:nil];

}

- (void)invokeEditor:(VPKImage*)image fromView:(UIView*)view {
    
    __block UIActivityIndicatorView* activityView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhite];
    [activityView stopAnimating];
    
    activityView.center = CGPointMake(self.editorPreview.bounds.size.width/2.0f, self.editorPreview.bounds.size.height/2.0f);

    dispatch_time_t  when  =dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC));
    dispatch_queue_t queue = dispatch_get_main_queue();
    dispatch_after(when, queue, ^{
        if (activityView) {
            [self.editorPreview addSubview:activityView];
            [activityView startAnimating];
        }
    });
    
    /*
        
     invoking the VPKVeepEditor
     
        
     set the editor's transitioning delegate to a custom transitioning object (or nil) to override supplied transition animations
     
     this code is all OPTIONAL - if you don't set the delegate on VPKPreview, presenting and dismissing behaviour occurs as a default. However, at some point before invoking the editor, user authentication needs to be dealt with.

    */
    __weak typeof(self) weakself = self;
    [VPKit authenticateWithEmail:@"test@example.com"
                      completion:^(BOOL success, NSInteger responseCode, NSError * _Nonnull error) {
         __strong typeof(self) strongself = weakself;
       
        /*
         
         Authentication 
         
         We authenticate before invoking the editor, as veep creation requires an authenticated user.
         
         Authenticated users are weak (no password) or strong (password-protected)
    
         If responseCode to a weak login attempt is 401 (unauthorised), we can make a similar call for strong authentication with a password.
        
         User account admin can be an implementation detail in the host app or the Veepio Developer control panel.
         
         */
                          [activityView removeFromSuperview];
                          activityView = nil;
        
        if (success) {
            strongself.vpEditor =  [VPKit editorWithImage:image
                                           fromView:view];
            strongself.vpEditor.delegate = self;
            strongself.vpEditor.modalPresentationStyle = UIModalPresentationOverFullScreen;
            [strongself presentViewController:strongself.vpEditor animated:YES completion:nil];
        } else {
            NSLog(@"%@",error);
        }
    }];
    

}


#pragma mark - VPKViewController delegate

/*
    example delegate methods for VPKVeepViewer
 
 */

- (void)veepViewer:(VPKVeepViewer *)viewer didFinishViewingWithInfo:(NSDictionary *)info {
    VPKPublicVeep* pVeep = info[@"veep"];
    NSLog(@"%s %@",__func__, pVeep);
    __weak typeof(self) weakself = self;
    [self dismissViewControllerAnimated:YES completion:^{
        [weakself.viewerPreview showIcon];
    }];
}

- (void)veepViewerDidCancel:(VPKVeepViewer *)viewer {
    NSLog(@"%s",__func__);
    [self dismissViewControllerAnimated:YES completion:nil];
}

/*
    example delegate mathods for VPKVeepEditor
 
 */

- (void)veepEditor:(VPKVeepEditor *)editor didPublishVeep:(NSString *)veepId  {
   
    NSLog(@"%s %@",__func__,veepId);
    [self dismissViewControllerAnimated:YES completion:^{
        //test that we can fetch the veep we just published
        [VPKit requestVeep:veepId completionBlock:
         ^(VPKPublicVeep * _Nullable veep, NSError * _Nullable error) {
             NSLog(@"%@",veep);
         }];
    }];
}

- (void)veepEditorDidCancel:(VPKVeepEditor *)editor {
    NSLog(@"%s",__func__);
    [self dismissViewControllerAnimated:YES completion:nil];

}





@end

File: Demo_Create_swift/VPKitDemoSwift/ViewController.swift
//
//  ViewController.swift
//  VPKitDemoSwift
//
//  Created by jonathan on 09/11/2016.
//  Copyright © 2016 Veepio. All rights reserved.
//

import UIKit
import VPKit

class ViewController:
    UIViewController
    , VPKVeepEditorDelegate
    , VPKVeepViewerDelegate
    , VPKPreviewDelegate
{
    
    //MARK: - properties
    
    @IBOutlet var titleLabel: UILabel!
    @IBOutlet var consumeLabel: UILabel!
    @IBOutlet var createLabel: UILabel!
    @IBOutlet var versionLabel: UILabel!
    @IBOutlet var viewerPreview: VPKPreview!
    @IBOutlet var editorPreview: VPKPreview!
    
    
    
    //MARK: - viewController lifecycle
    
    required init?(coder aDecoder: NSCoder)
    {
        super.init(coder: aDecoder)
        //fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad()
    {
        super.viewDidLoad()
        self.versionLabel.text = VPKit.version()
        self.configureViewerWithTestImage()
        self.configureEditor()
        self.configureConstraints()
        
    }
    
    

    //MARK: - configuration
    
    func configureViewerWithTestImage()
    {
        guard let image = UIImage.init(named: "KrispyGlas") else {return}
        let previewImage: VPKImage = VPKImage(image: image, veepID:"560")
        self.viewerPreview.image = previewImage
    }
    
    func configureViewerWithTestVideo()
    {
        guard let image = UIImage.init(named: "tomcruise") else {return}
        let previewImage: VPKImage = VPKImage(image: image, veepID:"680")
        self.viewerPreview.image = previewImage
    }
    
    func configureEditor()
    {
        guard let image = UIImage.init(named: "stock_photo") else {return}
        let previewImage: VPKImage = VPKImage(image: image, veepID:nil)
        self.editorPreview.image = previewImage
        /*
         for the editor example, we'll set an optional delegate.
         */
        self.editorPreview.delegate = self
    }
    
    
    
    
    //MARK: -  VPKPreview delegate
    
    func vpkPreviewTouched(_ preview:VPKPreview, image:VPKImage)
    {
        if preview.isEqual(self.viewerPreview) {
            self.invokeViewer(image: image, fromView: preview)
        } else {
            self.invokeEditor(image: image, fromView: preview)
        }
    }

    func invokeViewer(image:VPKImage, fromView view:UIView)
    {
        
        /*
         invoking the VPKVeepViewer
         
         set the viewer's transitioning delegate to a custom transitioning object (or nil) to override supplied transition animations
         
         this code is all OPTIONAL - if you don't set the delegate on VPKPreview, presenting and dismissing behaviour occurs as a default
         */
        
        
        guard let vpViewer = VPKit.viewer(with: image, from: view) else { return }
        vpViewer.delegate = self
        vpViewer.modalPresentationStyle = UIModalPresentationStyle.overFullScreen
        self.present(vpViewer, animated: true, completion: nil)
    }
    
    func invokeEditor(image:VPKImage, fromView view:UIView)
    {
        
        
        /*
         
         invoking the VPKVeepEditor
         
         
         set the editor's transitioning delegate to a custom transitioning object (or nil) to override supplied transition animations
         
         this code is all OPTIONAL - if you don't set the delegate on VPKPreview, presenting and dismissing behaviour occurs as a default. However, at some point before invoking the editor, user authentication needs to be dealt with.
         
         */
        
        
        VPKit.authenticate(withEmail: "public_test_2_4_6@example.com")
        {
            [weak self] (success, responseCode, error) in
            
            /*
             
             Authentication
             
             We authenticate before invoking the editor, as veep creation requires an authenticated user.
             
             Authenticated users are weak (no password) or strong (password-protected)
             
             If responseCode to a weak login attempt is 401 (unauthorised), we can make a similar call for strong authentication with a password.
             
             User account admin can be an implementation detail in the host app or the Veepio Developer control panel.
             
             */
            
            if (success) {
                guard let vpEditor = VPKit.editor(with: image, from: view) else { return }
                vpEditor.delegate = self
                vpEditor.modalPresentationStyle = UIModalPresentationStyle.overFullScreen
                self?.present(vpEditor, animated: true, completion: nil)

            } else {
                print("\(error)")
            }
        }
        
        
    }
    
    



    
    
    
    //MARK: - VPKViewController delegate functions
    
    /*
     delegate methods for VPKVeepViewer
     
     */
    
    func veepViewer(_ viewer: VPKVeepViewer, didFinishViewingWithInfo info: [AnyHashable : Any]?)
    {
        
        // the veep object can be read from info["veep"] if required

        self.dismiss(animated: true, completion:  {
            self.viewerPreview?.showIcon(true)
        })
    }
    
    func veepViewerDidCancel(_ viewer: VPKVeepViewer) {
        self.dismiss(animated: true, completion: nil)
    }
    
    /*
     delegate mathods for VPKVeepEditor
     
     */
    
    func veepEditor(_ editor: VPKVeepEditor, didPublishVeep veepID: String)
    {
        /*
         demo method to get a publicVeep object from a veepID
         */
        VPKit.requestVeep(veepID) { (veep, error) in
            if ((veep) != nil){
                print("\(veep)")
            } else {
                print("\(error)")
            }
        }
        self.dismiss(animated: true, completion: nil)
    }
    
    func veepEditorDidCancel(_ editor: VPKVeepEditor)
    {
        self.dismiss(animated: true, completion: nil)
    }
    
    
    //MARK: - layout boilerplate
    
    
    func configureConstraints()
    {
        let views:[String:UIView] =  [
            "titleLabel":self.titleLabel!
            ,"consumeLabel":self.consumeLabel!
            ,"createLabel":self.createLabel!
            ,"viewerPreview":self.viewerPreview!
            ,"editorPreview":self.editorPreview!
            ,"versionLabel":self.versionLabel!

            
        ]
        for (_,view) in views {
            view.translatesAutoresizingMaskIntoConstraints = false
        }
        
        var ratio : CGFloat =  1.0
        let height = self.view.bounds.size.height

        if (height>0) {
            let width = self.view.bounds.size.width
            ratio = width/height
        }
        
        
        
        let vMargin = 80 * ratio;
        let tMargin = height / 20.0
        let formats =
            [
                "H:|-(vMargin)-[titleLabel]-(vMargin)-|",
                "H:|-(vMargin)-[viewerPreview]-(vMargin)-|",
                "H:|-(vMargin)-[consumeLabel]-(vMargin)-|",
                "H:|-(vMargin)-[editorPreview]-(vMargin)-|",
                "H:|-(vMargin)-[createLabel]-(vMargin)-|",
                "H:|-(vMargin)-[versionLabel]-(vMargin)-|",
                "V:|-(tMargin)-[titleLabel]-(18)-[viewerPreview]-(4)-[consumeLabel]-(18)-[editorPreview]-(4)-[createLabel]",
                "V:[versionLabel]-(8)-|"
        ];
        
        for format in formats {
            self.view.addConstraints(
                NSLayoutConstraint.constraints(withVisualFormat: format,
                                               options:NSLayoutFormatOptions(rawValue:0),
                                               metrics: ["vMargin":vMargin, "tMargin":tMargin],
                                               views: views)
            )
        }
        
        guard let vImage = self.viewerPreview?.image else {return}
        var size = vImage.size;
        self.alignHeightToWidth(view:self.viewerPreview!, ratio: size.height/size.width);
        guard let eImage = self.editorPreview?.image else {return}
        size = eImage.size;
        self.alignHeightToWidth(view:self.editorPreview!, ratio: size.height/size.width);
        
    }
    
    func alignHeightToWidth (view:UIView, ratio:CGFloat)
    {
        view.superview?.addConstraint(
            NSLayoutConstraint.init(
                item: view,
                attribute: NSLayoutAttribute.height,
                relatedBy: NSLayoutRelation.equal,
                toItem: view,
                attribute: NSLayoutAttribute.width,
                multiplier: ratio, constant: 0)
        )
    }
    

}

